tree/ntuple/v7/inc/ROOT/RFieldBase.hxx

```diff
@@ -500,50 +360,17 @@ protected:
    /// column type exists.
    virtual std::size_t AppendImpl(const void *from);
    virtual void ReadGlobalImpl(NTupleSize_t globalIndex, void *to);
-   virtual void ReadInClusterImpl(RClusterIndex clusterIndex, void *to)
-   {
-      ReadGlobalImpl(fPrincipalColumn->GetGlobalIndex(clusterIndex), to);
-   }
+   virtual void ReadInClusterImpl(RClusterIndex clusterIndex, void *to);
 
    /// Write the given value into columns. The value object has to be of the same type as the field.
    /// Returns the number of uncompressed bytes written.
-   std::size_t Append(const void *from)
-   {
-      if (~fTraits & kTraitMappable)
-         return AppendImpl(from);
-
-      fPrincipalColumn->Append(from);
-      return fPrincipalColumn->GetElement()->GetPackedSize();
-   }
+   std::size_t Append(const void *from);
 
    /// Populate a single value with data from the field. The memory location pointed to by to needs to be of the
    /// fitting type. The fast path is conditioned by the field qualifying as simple, i.e. maps as-is
    /// to a single column and has no read callback.
-   void Read(NTupleSize_t globalIndex, void *to)
-   {
-      if (fIsSimple)
-         return (void)fPrincipalColumn->Read(globalIndex, to);
-
-      if (fTraits & kTraitMappable)
-         fPrincipalColumn->Read(globalIndex, to);
-      else
-         ReadGlobalImpl(globalIndex, to);
-      if (R__unlikely(!fReadCallbacks.empty()))
-         InvokeReadCallbacks(to);
-   }
-
-   void Read(RClusterIndex clusterIndex, void *to)
-   {
-      if (fIsSimple)
-         return (void)fPrincipalColumn->Read(clusterIndex, to);
-
-      if (fTraits & kTraitMappable)
-         fPrincipalColumn->Read(clusterIndex, to);
-      else
-         ReadInClusterImpl(clusterIndex, to);
-      if (R__unlikely(!fReadCallbacks.empty()))
-         InvokeReadCallbacks(to);
-   }
+   void Read(NTupleSize_t globalIndex, void *to);
+   void Read(RClusterIndex clusterIndex, void *to);
```

Comment:
Outlining `Append` and `Read` does not allow the compiler to inline them anymore. For simple fields, this inlining allows to directly call into the (principal) `RColumn`. I'm not sure if this is important, but it's also not clear to me why these functions have to be outlined? They don't seem to use any of the outlined `class`es / `struct`s...

Reply:
No reason other than "they take space". If it's important for them to be inlined for performance reasons then I'll just put them back there

Reply:
Good point! Yes, these two methods should remain inline-able.